Add unit test to ensure no gradients sync when calling ddp.module(input)

chenyangyu1988 opened the pull request:
Summary:
Add unit test to ensure no gradients sync when calling ddp.module(input), e.g not invoking prepare_for_backward

PyText is depending on DDP for data parallel distributed training. To support accumulate gradients locally before gradients sync, we are calling orig_model.forward instead of ddp_model.forward. Add a unit test to avoid changes break the assumption.

Differential Revision: D15263155

mrshenli commented on test/test_c10d.py:
```diff
@@ -2370,6 +2370,57 @@ def check_no_grads():
+            ddp_model.module(input)
+
+        # check two model parameters over 2 iterations
+        for iteration in range(2):
```
Let's do 4 iterations to make sure that things are still correct after the first sync.

mrshenli commented on test/test_c10d.py:
```diff
@@ -2370,6 +2370,57 @@ def check_no_grads():
 
+    @skip_if_not_multigpu
+    @skip_if_not_nccl
+    def test_accumulate_gradients(self):
```
Shall we add a comment to say that this would only work when this is a single model replica per process.

mrshenli reviewed (approved):
Thanks for adding the test. I feel this test should only temporarily live to make sure changes in PyTorch does not break your use cases. And it should retire when we have a better solution (e.g., a context manager as @apaszke suggested).

chenyangyu1988 committed:
Add unit test to ensure no gradients sync when calling ddp.module(input) (#20282)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20282

Add unit test to ensure no gradients sync when calling ddp.module(input), e.g not invoking prepare_for_backward

PyText is depending on DDP for data parallel distributed training. To support accumulate gradients locally before gradients sync, we are calling orig_model.forward instead of ddp_model.forward. Add a unit test to avoid changes break the assumption.

Reviewed By: mrshenli

Differential Revision: D15263155

fbshipit-source-id: f5bd884616b2783064f383820dc6397f8608c3c6
